### ISSUE TITLE
Variations in d-space

### ISSUE THREAD

**Issue by hgretar** [kklmn/xrt#45]
I'm interested in simulating the effect of thermal gradient in a crystal. What I have in mind is to control manually the variations of the d-spacing as a function of the local xyz OE coordinates. 

I know how to alter the shape of the OE but I have not been able to find a way to change the d-spacing of the material itself.  Do you think this is possible?

**Comment by kklmn** [comment]
The machinery is already there: we can do laterally graded multilayers. The same methods can be added to Crystal class.

**Comment by hgretar** [comment]
Great, that is even better.

**Comment by hgretar** [comment]
I've noticed that this is still an open request. Is this more of a work than you initially thought?

**Comment by kklmn** [comment]
If you have relevant experimental data that you can share, I can quickly implement it. Otherwise, you will have to program it yourself, I can guide you about the needed modifications.

**Comment by hgretar** [comment]
We have some preliminary results for the following analyzer setup in the (+,+) geometry:

Incoming beam: 2840+/- 0.015 eV
Incoming divergence: +/-50urad
Incoming beam size: 6x6mm2

First reflection: Ge(111) alpha=38.8deg, Bragg 41.9 degrees -> incidence angle around 3.1 degrees.
Second reflection:  Ge(111) alpha=-38.8deg

In our setup we rock the second Ge(111) and look at the angular width. Theory (XRT) tells us that the width should be about 85urad and is fairly insensitive to the incoming beam collimation or energy bandwidth.  

What we find instead is a width of about 130urad that we believe is coming from variations in d-spacing along the crystal surfaces (footprint on the xtals is 100mm), either due to thermal gradient or simply intrinsic.  

I'll try to modify the code to mimic this but if this is something of interest to you I would be happy if you could have a look too.

**Comment by kklmn** [comment]
How thick are the crystals? Do you know about any surface treatment? Can you try with the triple energy?

**Comment by hgretar** [comment]
The crystals have a dimension  of 150x15x10mm3 (LxWxH, 10mm thick) and where polished (5micron). We will try the 5x energy next month and do proper topography on their surface. That profile could then possibly be included in our simulations.

**Comment by kklmn** [comment]
So no chemical etching after polishing? Might be the surface amorphous layer is the culprit of the broadening...

**Comment by hgretar** [comment]
Correct no chemical etching. Our data does however show that if you simply select a smaller footprint (10mm instead of 100mm) on the Ge crystals the width is 85urad, the problem is that the center varies across the Ge surface which brings me back to d-space variations on a macro scale.

**Comment by kklmn** [comment]
For a larger footprint, not only d may vary but also the figure. I suppose you've already tried a curved surface.

**Comment by hgretar** [comment]
yes curved I 've tried and it can easily cause this. This is in principle can be fixed but I'd like to exclude the d-space variations.

**Comment by kklmn** [comment]
Did you work at room T? Do you have data on thermal expansivity of Ge?

**Comment by hgretar** [comment]
Yes room temperature. In these conditions the thermal expansion of Ge is 5.9 x 10-6 K-1, about double the one for Silicon.

**Comment by kklmn** [comment]
I will start working on it in a week.

**Comment by hgretar** [comment]
No problem and thanks for your help

**Comment by kklmn** [comment]
Hi, I've done the modifications, they were very few. And in my tests, it works. Please try the GH version.

You will need to define `get_d()` in you crystal class, like this:
```
class Crystal_d(rm.CrystalSi):
    def get_d(self, x, y):
        return self.d * (1+y*1e-5)
```

**Comment by hgretar** [comment]
I've tried it and it seems to work for the reflected intensity but it has no affect on the reflected angle. In principle if the d-space variations are fixed (i.e. do not depend on y-coordinate) I should be able to get the same results if I simply change the lattice constant a. This is where things get interesting:

1) When I use Ge(111) with alpha=0 these two approaches give the same reflected beam.
2) When I use Ge(111) with alpha=-38.8degrees (incidence angle is 3degrees) these two approaches give different reflected beam (they are off in angle)

So changing d-spacing gives a different results when using asymmetric crystal compared to changing only the lattice constant. Not sure if that is correct.

**Comment by kklmn** [comment]
You're right, in an asymmetric crystal, the departure angle depends on d. It's great that you've noticed it. I've now corrected that part as well. Please try it.

**Comment by hgretar** [comment]
That was very quick, thanks a lot. Results look consistent now :)

**Comment by kklmn** [comment]
Great!
